openshift-kube-apiserver/admission/customresourcevalidation/dns/validate_dns.go

```diff
@@ -142,3 +145,109 @@ func validateTolerations(versionedTolerations []corev1.Toleration, fldPath *fiel
 	allErrors = append(allErrors, apivalidation.ValidateTolerations(unversionedTolerations, fldPath)...)
 	return allErrors
 }
+
+// validateUpstreamResolversCreate validates configuration of the Upstream objects when TLS is configured.
+func validateUpstreamResolversCreate(upstreamResolvers operatorv1.UpstreamResolvers, fieldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+
+	errs = append(errs, validateDNSTransportConfig(upstreamResolvers.TransportConfig, fieldPath)...)
+
+	if upstreamResolvers.TransportConfig.Transport == operatorv1.TLSTransport {
+		// Transport is TLS so we must check if there are mixed Upstream types. SystemResolveConf is not allowed with TLS.
+		for i, upstream := range upstreamResolvers.Upstreams {
+			errPath := fieldPath.Child("upstreams")
+			if upstream.Type == operatorv1.SystemResolveConfType {
+				errMessage := "SystemResolveConf is not allowed when TLS is configured as the transport"
+				errs = append(errs, field.Invalid(errPath.Index(i).Child("type"), upstream.Type, errMessage))
+			}
+		}
+	}
+
+	return errs
+}
+
+// validateUpstreamResolversUpdate validates configuration of the Upstream objects when TLS is configured.
+func validateUpstreamResolversUpdate(newUpstreamResolvers operatorv1.UpstreamResolvers, oldUpstreamResolvers operatorv1.UpstreamResolvers, fieldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	newUpstreams := newUpstreamResolvers.Upstreams
+	oldUpstreams := oldUpstreamResolvers.Upstreams
+	newTransport := newUpstreamResolvers.TransportConfig.Transport
+
+	// Upstreams has changed and the Transport is known
+	if !reflect.DeepEqual(newUpstreams, oldUpstreams) && isKnownTransport(newTransport) {
+		errs = append(errs, validateUpstreamResolversCreate(newUpstreamResolvers, fieldPath)...)
+	}
+
+	// TransportConfig has changed
+	if !reflect.DeepEqual(newUpstreamResolvers.TransportConfig, oldUpstreamResolvers.TransportConfig) {
```

Comment:
`validateUpstreamResolversCreate` calls `validateDNSTransportConfig`, so you could end up doing the same validation twice.
```suggestion
	} else if !reflect.DeepEqual(newUpstreamResolvers.TransportConfig, oldUpstreamResolvers.TransportConfig) {
		// TransportConfig has changed
```
Alternatively, the logic might be clearer and issues like this more obvious if you factored the upstreams check into a function that both `validateUpstreamResolversCreate` and `validateUpstreamResolversUpdate` could use.

Reply:
I went with the `else if` option (fixed locally) so as to not hold things up but I think refactoring out that logic is a good idea; I can address that separately.